Add easy part of critical-1 task

malaartyom opened the pull request.

malaartyom committed:
Add easy part of critical-1 task

malaartyom committed:
Add Medium part of critical-1 task.

Now I'am using ReadWriteLock instead of syncronization on Object lock
Level two compilation of methods happends in other thread. For this purpose I use
ExecutorService.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
-    // TODO: add methods
-    // TODO: add inner classes
-    // TODO: add utility classes in the same package
+// TODO: add methods
```
Тудушки еще релевантны?

malaartyom replied:
Нет, не релевантны. Убрал.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
+     * Stores the usage count of methods, where the key is the method's {@code id}
+     * and the value is the number of times it has been used.
+     */
+    private static final HashMap<Long, Integer> usages = new HashMap<>();
```
Стоит вынести общее состояние в другой класс. Данный класс становится перегруженным.

malaartyom replied:
Вынес в отдельный класс BalancerState. Добавил соответствующих методов для обновления состояния полей и получения из этих полей информации

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
+     * @param id the {@link MethodID} whose usage count should be incremented
+     */
+    private void incUsages(MethodID id) {
+        lock.writeLock().lock();
```
Изменение `synchronized` на `RWLock` вам особо не помог. Проблема в том, что для такой простой операции, как инкрементирование счетчика, возникает взаимодействие с другими потоками.

Нужно сделать частые операции:
- либо потоково-локальными
- либо read-only

Это позволит избежать проблем с производительностью.

Например, счетчик стоит сделать потоково-локальным. Это приведет к появлению новых проблем. Например, так просто гарантировать условие `Weak-global-caching` из 1 уровня не получится. Их нужно тоже решить.

malaartyom replied:
А это обязательнее условие для сдачи на middle и easy? Пока не представляю, как сделать счётчик локальным и при этом избежать появления новых проблем.

rzapanov replied:
Разве `easy` часть не была засчитана очно в пятницу? `easy` у вас засчитан.

Для`medium` я требую это условие. Целью задачи (и большинства других задач на курсе) является не только написать корректно работающую программу, но и с разумной производительностью. Прочувствовать различные аспекты, конечно, в рамках имеющихся тестов особо не получится, но какие-то аспекты прочувствовать нужно.

Одна из самых частых проблем многопоточности - излишняя синхронизация. При полностью параллельном исполнении у независимых агентов (потоков, ядер или даже разных узлов в кластере) обычно нет причин простаивать, но при взаимодействии друг с другом требуется каким-нибудь способом "договориться" между агентами, например, о том кто кто из потоков сумел взять во владение блокировку. При частом взаимодействии агентов проблемы с производительностью становяться заметны.

Утверждается, что взятие блокировки на каждую процедуру "вызова" метода - достаточно дорого.

Подсказка: попробуйте выполнять профилирование (подсчет вызовов) локально, выполняя раз в `K` профилирований синхронизацию с глобальным кэшем. Рассуждайте о `Weak-global-caching` не как о строгом условии, а рассмотрите достаточно большой класс "исполнений", где оно выполняется. Например, если количество потоков ограничено числом `N`.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
+            if (usages.get(id.id()) >= L1) {
+                if (usages.get(id.id()) >= L2 && !compiledL2.contains(id.id())) {
+                    compiledL2.add(id.id());
+                    executor.submit(() ->
```
Возможна следующая ситуация:
1. поток отправил на компиляцию задачу
2. вышел из метода `check`
3. увидит в таблице `methods` старый `l1` метод.
4. при повторном исполнении метода не гарантируется, что там появится `l2` метод, если компиляция для `l2` выполняется дольше 100000 запусков метода.

Из-за чего свойство `Eventual-per-thread-progress-2` не гарантируется

malaartyom replied:
Добавил HashMap compiling для отслеживания тех методов, которые слишком долго компилируются под l2.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
+    /**
+     * Contains the {@code id} of methods that have been compiled at level 2 (L2).
+     */
+    private static final HashSet<Long> compiledL2 = new HashSet<>();
```
Наименование смущающее. 
Таблица называется `compiledL2`, но для проверки одинаково названного предиката `isCompiled` используется таблица `methods`.
Дока к таблицам не соответсвует реальному поведению: возможна ситуация, когда `compiledL2(id) == true`, но в `methods` метод еще не появился.

malaartyom replied:
Хорошо, переименовал compiledL2 -> compiledByLevel2 и methods -> compiledMethods. 

Кажется для проверки предиката isCompiled всё же нужно обращаться к полю compiledMethods. Ведь в этой таблице мы храним скомпилированные методы вне зависимости от уровня. Также я изменил немного метод check, теперь, кажется, невозможна ситуация, когда compiledByLevel2(id) == true, но в compiledMethods метод еще не появился.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -3,26 +3,149 @@
+     * @param id
+     * @return {@link CompiledMethod}
+     */
+    private CompiledMethod getCompiled(MethodID id) {
```
Для взятия элемента из таблицы `methods` требуется сначала проверить `isCompiled`, который требует взятия лока, а потом лок берется здесь. Попробуйте объединить это действие в одно. Возвращаемым значением удобно использовать `Optional`.

malaartyom replied:
Убрал isCompiled, теперь всё делаю внутри getCompiled и возвращаю Optional. Пока закомментил isCompiled, потому что не уверен, что вы именно этого хотели.

rzapanov replied:
Примерно это и ожидалось. Правда там новые проблемы появились, связанные с `updateCompiling`

rzapanov reviewed (changes requested):
Для L2 стоит код немного оформить, выделив хотябы 1-2 дополнительных класса.
Помимо замечания с `Eventual-per-thread-progress-2` требуется уменшить количество синхронизации.

malaartyom committed:
Imporve porblems and add additional functionality.

Added Eventual-per-thread-progress-2 constraint, so now all the requirenments
for middle part are completed (I hope so)

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -1,28 +1,73 @@
+     * A thread-safe pool of threads for compiling methods asynchronously.
+     * <p>
+     */
+    private final ExecutorService executor = Executors.newCachedThreadPool();
```
Раньше здесь был `static` и экзекутор разделялся между потоками.
Теперь у каждого пользовательского потока будет свой `executor`. Т. к. потоки - дорогой и ограченный ресурс, то забирать их у "пользователя" в таком количестве - плохо.

P. S. `static` оставлять тоже не стоит, т.к. объект будет, кажется, разделяться между разными запусками тестов. Стоит заранее создать его вместе с `TestEnvironment` и передать его в конструктор `UserThread`.

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/SolutionThread.java:
```diff
@@ -1,28 +1,73 @@
+        state.incrementUsages(id);
+        check(id);
+        if (state.checkTimeOfCompilation(id)) {
+            executor.shutdown();
```
Почему `shutdown`?

malaartyom replied:
А как по-другому?

rzapanov commented on hw/block1/critical/jit-balancer/src/main/java/org/nsu/syspro/parprog/solution/BalancerState.java:
```diff
@@ -0,0 +1,235 @@
+        lock.readLock().lock();
+        try {
+            if (compiling.containsKey(id.id())) {
+                updateCompiling(id);
```
Модификация под `readLock` приведет к гонке.

malaartyom replied:
Поменял на writeLock

malaartyom committed:
Improve solution

malaartyom committed:
Make some changes

Imporved:
 + Usage of future instead of executor.shotdown()
 + BalancerState now in constructor
 + Finally there are two counters: local and global
